### DISCUSSION TITLE
set_interrupt_handler example usage

### DISCUSSION THREAD

**Discussion by SreeniIO** [HiRoFa/quickjs_es_runtime#49] [Q&A] [answered]
Hi there!,

I'm trying to explore the usage of set_interrupt_handler to prevent long running or runaway functions (e.g. infinite loop etc.), but the handler is not getting called. Am I missing something? Below is the sample code where the registered interrupt handler never gets called.

```rust
use futures::executor::block_on;
use log::LevelFilter;
use quickjs_runtime::esruntime::EsRuntime;
use quickjs_runtime::esruntimebuilder::EsRuntimeBuilder;
use quickjs_runtime::esscript::EsScript;
use quickjs_runtime::esvalue::*;
use std::sync::Arc;
use std::{thread, time};

async fn test(rt: Arc<EsRuntime>) {
    let _res = rt
        .eval(EsScript::new(
            "basics.es",
            r#"
                for (let i = 0; i < 100; i++) {
                    rust.sleep();
                    console.log(i);
                }
            "#,
        ))
        .await
        .ok()
        .unwrap();
}

fn main() {
    simple_logging::log_to_stderr(LevelFilter::Info);
    let mut rt = EsRuntimeBuilder::new();
    rt.set_interrupt_handler(|_| {
        // this handler never gets invoked
        println!("in interrupt handler...");
        true
    });
    let rt = rt.build();

    rt.set_function(vec!["rust"], "sleep", |_, _| {
        let one_sec = time::Duration::from_secs(1);
        thread::sleep(one_sec);
        Ok(1.to_es_value_facade())
    })
    .ok()
    .expect("set sleep function failed");

    block_on(test(rt));
}
```

**Comment by andrieshiemstra** [comment, marked as answer]
Hmm yes sorry.. I added the interrupt handler without testing it or adding a test-case

The implementation actually did nothing, I fixed it in the main branch now.

I'm not sure if it will actually work with your code, as far as I can tell the implementation in quickjs is based on calling it every x actions in quickjs, doing a sleep in rust will not add to that count and thus the interrupt handler may not be called.

In my test case I do a for loop with 10k calls and the interrupt handler is called about 4 times.

Hope this helps,

Kind regards.

**Reply by SreeniIO** [reply]
Wow that was quick. Thanks. Any ETA on when version 0.5 will be published?

**Reply by andrieshiemstra** [reply]
Well i wanted tot publish 5.0 with full support for fetch with all its options if i can't finish that before my vacation i'll see about publishing it as-is somewhere in the next two weeks

**Reply by SreeniIO** [reply]
Hi again,
The interrupt_handler is getting called as expected, but how do I know for which script it's getting invoked? It would help if I can access the current context in which the script is running. Or at least the duration of how long the current script has been running.

```rust
let ctx_id: String = ...some_id;
rt.create_context(&ctx_id);

rt.add_rt_task_to_event_loop(move |q_js_rt| {
  let q_ctx = q_js_rt.get_context(&ctx_id);
  q_ctx.eval(Script....); // here if the interrupt_handler is invoked, is there a way to access the above ctx_id
   ...
});
```

Thanks!

**Reply by andrieshiemstra** [reply]
Hi, you can use ```quickjs_utils::get_script_or_module_name_q```

> ps, based on your example i think i'll modify the handler to include the context

scratch that, the native variant does not pass a context

**Reply by andrieshiemstra** [reply]
https://github.com/bellard/quickjs/issues/74

i asked if this could be altered, if it is not possible i could allways keep the current ctx in a thread_local

**Reply by andrieshiemstra** [reply]
funny :( , it isn't bulletproof
```
let rt = EsRuntimeBuilder::new()
            .set_interrupt_handler(move |qjs_rt| {
                let q_ctx = qjs_rt.get_main_context();
                let script_name = get_script_or_module_name_q(q_ctx).ok().expect("poof");
                println!("ihandler called @ {}", script_name);
                false
            })
            .build();
        match rt.eval_sync(Script::new(
            "test_interrupt.es",
            "for (let x = 0; x < 10000; x++) {console.log('x' + x);}",
        )) {
            Ok(_) => {}
            Err(err) => {
                panic!("err: {}", err);
            }
        }
```
outputs:
```
ihandler called @ 
ihandler called @ test_interrupt.es
ihandler called @ test_interrupt.es
ihandler called @ test_interrupt.es
```

**Reply by SreeniIO** [reply]
It would be good to get hold of the context id instead of the script name. As the same script can be executed by multiple user sessions concurrently.

**Reply by SreeniIO** [reply]
Also in my case, I don't use the main context to execute the scripts. Hence getting an empty string with the above code sample.

**Reply by andrieshiemstra** [reply]
Ha, if you check the code for getModuleOrScriptName at https://raw.githubusercontent.com/bellard/quickjs/master/quickjs.c

You'll see that the context is only used to get the runtime.. so you can pass any context and thus can simply allways use the q_js_rt.get_main_context() to get the script name even if you're running code in a different context

So that's the good

The bad is that get_script_or_module_name_q sometimes panics (even when using correct ctx) i'm looking into that...

**Reply by andrieshiemstra** [reply]
> The bad is that get_script_or_module_name_q sometimes panics (even when using correct ctx) i'm looking into that...

was due to some quickjs internal (filename was <brand> which reports as len 0) the assertions was added when is was clueless about dealing with Cstr, so i removed it now

**Reply by SreeniIO** [reply]
Yeah there seems to be something wrong with the way the filename is retrieved. Once I saw my DB_URL value got printed!!! I do see some random string getting returned by that function some times. But when a infinite loop runs (which is what I'm more interested to interrupt) it always returns the right file name and I've included the context id as part of the file name to resolve my use case. Based on the context id I can now retrieve the start time and the number of times the interrupt handler got called and that's a good information for me to decide when to terminate the script execution :)

Thanks.

**Reply by andrieshiemstra** [reply]
> Once I saw my DB_URL value got printed!!!

that should be fixed now (https://github.com/HiRoFa/quickjs_es_runtime/commit/fc1a752e1e759e2a1810999f48d8467ce83d401a)
